exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/diagnosiskey/DiagnosisKeyDao.java

```diff
@@ -119,31 +122,59 @@ public boolean verify(int verificationId, String requestChecksum, long totalKeyC
         }
     }
 
-    public List<TemporaryExposureKey> fetchAvailableKeysForEfgs(long operationId) {
+    @Transactional
+    public Optional<FederationOutboundOperation> fetchAvailableKeysForEfgs(boolean retry) {
         LOG.info("Fetching queued keys not sent to efgs.");
-        String sql = "select key_data, rolling_period, rolling_start_interval_number, transmission_risk_level, " +
-                "visited_countries, days_since_onset_of_symptoms, origin, consent_to_share " +
+        Timestamp timestamp = new Timestamp(Instant.now().toEpochMilli());
+        String sql = "with batch as ( " +
```

Comment:
Yep, that was on my mind also. I changed it to repeatable isolation level for now, which should do the trick if both nodes tries to do this concurrently, because we don't mind that other one just fails when the changed data has been committed. Anyway, let's discuss about this more if needed.

Reply:
And, put second thought for this. Repeatable read is probably too restrictive in this. I added for update skip locks and order by for select. Let's see how this ends.